errors/errors.go

```diff
@@ -15,209 +15,389 @@ package errors
+	return b
+}
+
+// WithHelpLink is used to pass HelpLink error details to the Error struct.
+func (b *ErrorBuilder) WithHelpLink(url string, description string) *ErrorBuilder {
+	link := errdetails.Help_Link{
+		Description: description,
+		Url:         url,
 	}
+	var links []*errdetails.Help_Link
+	links = append(links, &link)
+
+	help := &errdetails.Help{Links: links}
+	b.err.details = append(b.err.details, help)
 
-	return ste
+	return b
 }
 
-// *** HTTP Methods ***
+// WithHelp is used to pass Help error details to the Error struct.
+func (b *ErrorBuilder) WithHelp(links []*errdetails.Help_Link) *ErrorBuilder {
+	b.err.details = append(b.err.details, &errdetails.Help{Links: links})
 
-// ToHTTP transforms the supplied error into
-// a GRPC Status and then Marshals it to JSON.
-// It assumes if the supplied error is of type Error.
-// Otherwise, returns the original error.
-func (e *Error) ToHTTP() (int, []byte) {
-	resp, err := protojson.Marshal(e.GRPCStatus().Proto())
-	if err != nil {
-		errJSON, _ := json.Marshal(fmt.Sprintf("failed to encode proto to JSON: %v", err))
-		return http.StatusInternalServerError, errJSON
+	return b
+}
+
+// WithErrorInfo adds error information to the Error struct.
+func (b *ErrorBuilder) WithErrorInfo(reason string, metadata map[string]string) *ErrorBuilder {
+	errorInfo := &errdetails.ErrorInfo{
+		Domain:   ErrMsgDomain,
+		Reason:   reason,
+		Metadata: metadata,
 	}
+	b.err.details = append(b.err.details, errorInfo)
 
-	return e.httpCode, resp
+	return b
 }
 
-// HTTPCode returns the value of the HTTPCode property.
-func (e *Error) HTTPCode() int {
-	if e == nil {
-		return http.StatusOK
+// WithFieldViolation is used to pass FieldViolation error details to the Error struct.
+func (b *ErrorBuilder) WithFieldViolation(fieldName string, msg string) *ErrorBuilder {
+	br := &errdetails.BadRequest{
+		FieldViolations: []*errdetails.BadRequest_FieldViolation{{
+			Field:       fieldName,
+			Description: msg,
+		}},
 	}
 
-	return e.httpCode
+	b.err.details = append(b.err.details, br)
+
+	return b
 }
 
-// JSONErrorValue implements the errorResponseValue interface (used by `github.com/dapr/dapr/pkg/http`).
-func (e *Error) JSONErrorValue() []byte {
-	b, err := protojson.Marshal(e.GRPCStatus().Proto())
-	if err != nil {
-		errJSON, _ := json.Marshal(fmt.Sprintf("failed to encode proto to JSON: %v", err))
-		return errJSON
-	}
+// WithDetails is used to pass any error details to the Error struct.
+func (b *ErrorBuilder) WithDetails(details ...proto.Message) *ErrorBuilder {
+	b.err.details = append(b.err.details, details...)
+
 	return b
 }
+
+// Build builds our error
+func (b *ErrorBuilder) Build() error {
+	// Check for ErrorInfo, since its required per the proposal
+	containsErrorInfo := false
+	for _, detail := range b.err.details {
+		if _, ok := detail.(*errdetails.ErrorInfo); ok {
+			containsErrorInfo = true
+			break
+		}
+	}
+
+	if !containsErrorInfo || len(b.err.details) == 0 {
```

Comment:
Nit, remove redundant check
```suggestion
	if !containsErrorInfo {
```